pkg/common/cns-lib/volume/listview.go

```diff
@@ -194,6 +202,21 @@ func (l *ListViewImpl) isClientValid() error {
 		return nil
 	}
 
+	log.Infof("current session is either nil or not authenticated")
+
+	if !reconnect {
+		log.Infof("logging out existing VC session...")
+		err := l.LogoutSession(l.ctx)
```

Comment:
if vc session is already logged, and l.LogoutSession(l.ctx) returns error due to already logged out, then we will have issue.

Reply:
let me check

Reply:
I referred this govmomi test https://github.com/vmware/govmomi/blob/759d9a893ad4bac61e5060e179646e4ae233239d/session/manager_test.go#L52-L73.

Based on the test, if we tried calling Logout() on a session which is already logged out, we would get `NotAuthenticated` fault.

So, one solution is to use `context.CancelFunc`.

https://github.com/vmware/govmomi/blob/759d9a893ad4bac61e5060e179646e4ae233239d/property/collector.go#L273-L279 WaitForUpdates internally has a check to stop listening if the context is cancelled.

We can maintain a separate context and CancelFunc only for this purpose. We’ll continue to use the context used in logging. This separate context will only be used for cancelling the WaitForUpdates method.
I’ve written a small script here -> https://go.dev/play/p/BOg5UBYxRFr

Reply:
I also tested on a wcp setup https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/2788#issuecomment-1935044879